emails/views.py

```diff
@@ -307,23 +307,22 @@ def _index_DELETE(request_data, user_profile):
 @csrf_exempt
 def sns_inbound(request):
     incr_if_enabled("sns_inbound", 1)
-    # We can check for some invalid values in headers before processing body
-    # Grabs message information for validation
-    topic_arn = request.headers.get("X-Amz-Sns-Topic-Arn", None)
-    message_type = request.headers.get("X-Amz-Sns-Message-Type", None)
+    # First thing we do is verify the signature
+    json_body = json.loads(request.body)
+    verified_json_body = verify_from_sns(json_body)
 
-    # Validates header
-    error_details = validate_sns_header(topic_arn, message_type)
+    # Validate ARN and message type
+    topic_arn = verified_json_body.get("TopicArn", None)
+    message_type = verified_json_body.get("Type", None)
+    error_details = validate_sns_arn_and_type(topic_arn, message_type)
     if error_details:
-        logger.error("validate_sns_header_error", extra=error_details)
+        logger.error("validate_sns_arn_and_type_error", extra=error_details)
         return HttpResponse(error_details["error"], status=400)
 
-    json_body = json.loads(request.body)
-    verified_json_body = verify_from_sns(json_body)
     return _sns_inbound_logic(topic_arn, message_type, verified_json_body)
 
 
-def validate_sns_header(topic_arn, message_type):
+def validate_sns_arn_and_type(topic_arn, message_type):
```

Comment:
Nit: Another opportunity to add some typing. The function signature will be easier than the return type:
```suggestion
def validate_sns_arn_and_type(topic_arn: Optional[str], message_type: Optional[str]) -> Optional[dict[str, Union[str, list[str]]]]
```

Reply:
I haven't found a technique I like for prioritizing when and how to add types. And before you chalk it up to my laziness ...

I wanted to see how the function understandability improved with these type declarations. It changed the intelli-sense pop-up from this:

<img width="892" alt="image" src="https://user-images.githubusercontent.com/71928/186800928-cec547f6-be22-4a56-bb69-df8ad52af454.png">

to this:
<img width="793" alt="image" src="https://user-images.githubusercontent.com/71928/186801012-f5531952-73ff-48ca-87cf-97e24eacac86.png">

Hot-takes:

* I *really* like the type hints on the function parameters. `topic_arn: str, message_type: str` is way more helpful than `topic_arn: Unknown, message_type: Unknown`
* I *really* like doc-blocks that explain the parameters & return values of the function. 
* Like you say - the complex return type is harder, and I'm not sure if `dict [str, str | list[str]] | None` is more helpful than just `dict [str, Unknown] | None`? In either case, I'm likely going to have to inspect the function to really know how to handle what it returns.

On that last point - for this particular function, it seems like `pywright` was able to automatically determine `dict [str, ...] | None` part of the return type, but then it couldn't determine the types of the values of the dictionary? And yet, when I hover over each of the values in the returned dictionary, `pywright` **IS** able to determine each individual type? The possible value types are:

* `error: str | None`
* `received_topic_arn: str`
* `supported_topic_arn: list[SupportsRichComparisonT@sorted | _T@sorted]`
* `received_sns_type: str`
* `supported_sns_types: list[str]`

So, I tried creating a `dataclass` for it:
```python
@dataclass
class SNSValidationErrorDetails:
    error: str
    received_topic_arn: str
    supported_topic_arn: list
    received_sns_type: str
    supported_sns_types: list[str]


def validate_sns_arn_and_type(topic_arn: str, message_type: str) -> Optional[SNSValidationErrorDetails]:
```

But that made the intellisense even **less** helpful (IMO):

<img width="772" alt="image" src="https://user-images.githubusercontent.com/71928/186804280-272f7e63-0178-482a-b651-8658a88b9a50.png">

Just based on this example, I think I like:

* Add type hints to function parameters
* Add doc blocks explaining the parameters & return values of functions
* Add **basic** return types to functions

So what I ended up with is:
```python
def validate_sns_arn_and_type(topic_arn: str, message_type: str) -> Optional[dict[str, Any]]:
```

Which generates this IntelliSense:

<img width="772" alt="image" src="https://user-images.githubusercontent.com/71928/186805551-3389ed26-8142-4afb-860f-7946fba3788a.png">